index.js

```diff
@@ -359,6 +406,50 @@ module.exports = function AsyncAppendOnlyLog(filename, opts) {
     })
   }
 
+  function overwrite(blockIndex, blockBuf, cb) {
+    cache.set(blockIndex, blockBuf)
+    const blockStart = blockIndex * blockSize
+    writeWithFSync(blockStart, blockBuf, null, cb)
+  }
+
+  function truncate(newLatestBlockIndex, cb) {
+    if (newLatestBlockIndex >= latestBlockIndex) return cb(null, 0)
+    const size = (latestBlockIndex + 1) * blockSize
+    const newSize = (newLatestBlockIndex + 1) * blockSize
+    for (let i = newLatestBlockIndex + 1; i < latestBlockIndex; ++i) {
+      cache.remove(i)
+    }
+    truncateWithFSync(newSize, function onTruncateWithFSyncDone(err) {
+      if (err) return cb(err)
+      const blockStart = newSize - blockSize
+      loadLatestBlock(blockStart, function onTruncateLoadedLatestBlock(err) {
+        if (err) return cb(err)
+        const sizeDiff = size - newSize
+        cb(null, sizeDiff)
+      })
+    })
+  }
+
+  function compact(cb) {
+    if (compaction) {
+      debug('compaction already in progress')
+      waitingCompaction.push(cb)
```

Comment:
I think this should error. Otherwise you can only see that it didn't do anything from the debug.

Reply:
Hmm, but the callback will be called when the previous compaction is done, so to the person calling this API, it would seem like it did compact. I'm worried that sending an error would make the usage of the API harder.

Reply:
Ah, and while compaction is running we block all changes? So in a sense, those it did "complete". In that case I think this is ok :)